### PR TITLE
Save and report cumulative processing time in Progress structure

### PR DESCRIPTION
**Pull request by traceon**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category:
- Other

Changelog entry:
- None

Detailed description:
- Add `cumulative_processing_time` in `Progress` and populate it where it is maintained

### CODE REVIEW

File: src/Core/Defines.h [hunk 1]
```diff
@@ -63,6 +63,7 @@
 
 #define DBMS_MIN_REVISION_WITH_LOW_CARDINALITY_TYPE 54405
 #define DBMS_MIN_REVISION_WITH_CLIENT_WRITE_INFO 54420
+#define DBMS_MIN_REVISION_WITH_CLIENT_ELAPSED_TIME 54444
```
Comment (vdimir) [inline src/Core/Defines.h:66]: Should we bump `DBMS_TCP_PROTOCOL_VERSION`?

**Review by vdimir — COMMENTED** [review]
LGTM

**Comment by vdimir** [comment]
Also could you tell why do you need elapsed_time in progress?

File: src/Core/Defines.h (hunk 1 above)
Comment (traceon) [reply src/Core/Defines.h:66]: Exactly! Done. Thanks.

File: src/Interpreters/executeQuery.cpp [hunk 2]
```diff
@@ -695,7 +695,7 @@ static std::tuple<ASTPtr, BlockIO> executeQueryImpl(
                 auto progress_callback = context.getProgressCallback();
 
                 if (progress_callback)
-                    progress_callback(Progress(WriteProgress(info.written_rows, info.written_bytes)));
+                    progress_callback(Progress(WriteProgress(info.written_rows, info.written_bytes), info.elapsed_seconds * 1'000'000'000 /* seconds to nanoseconds */));
```
[the remark this answers is not in the document]
Comment (traceon) [reply src/Interpreters/executeQuery.cpp:698]: Thanks.

**Comment by traceon** [comment]
Originally, we needed elapsed time in `X-ClickHouse-Progress` HTTP headers for our platform (at Altinity.) It could be probably grafted in an ad-hoc manner somewhere around that code, but that wouldn't be the right way to do it.

@alex-zaitsev may provide more details for when we needed this.

**Comment by alexey-milovidov** [comment]
Why don't measure elapsed time on client (clickhouse-client does it to calculate speed)?
Another question - how elapsed times are merged from multiple servers in distributed query execution?

**Comment by traceon** [comment]
Also, please note the following somewhat relevant places, where I decided not to update the code to read/write elapsed time:

1. This uses its own stopwatch to measure things, it covers a bit more than would be accumulated in progress.elapsed time, so not sure if I should change this:
https://github.com/ClickHouse/ClickHouse/blob/d8b9278193405512bc96bd74d3f30d281c481bb9/programs/benchmark/Benchmark.cpp#L404-L430
2. In client it meausres elapsed time manually, and doesn't show it during progress, so I left it as-is:
https://github.com/ClickHouse/ClickHouse/blob/d8b9278193405512bc96bd74d3f30d281c481bb9/programs/client/Client.cpp#L1210-L1322
and
https://github.com/ClickHouse/ClickHouse/blob/d8b9278193405512bc96bd74d3f30d281c481bb9/programs/client/Client.cpp#L1967-L2069
3. Not sure if adding stopwatch here and measuring elapsed time makes sense and could be done accurately at all:
https://github.com/ClickHouse/ClickHouse/blob/d8b9278193405512bc96bd74d3f30d281c481bb9/src/DataStreams/CountingBlockOutputStream.cpp#L15-L30

**Comment by traceon** [comment]
>  Another question - how elapsed times are merged from multiple servers in distributed query execution?
@alexey-milovidov if they send multiple Progress packages, which are processed in same place, the elapsed time increments will be aggregated together in the common Progress snapshot. That's just a wild guess...

**Comment by traceon** [comment]
> Why don't measure elapsed time on client (clickhouse-client does it to calculate speed)?

Re clickhouse-client - yes, and I questioned/mentioned that above.
Re why we don't do it at client side - @alex-zaitsev could you please comment on this?

**Comment by alex-zaitsev** [comment]
> > Why don't measure elapsed time on client (clickhouse-client does it to calculate speed)?
> 
> Re clickhouse-client - yes, and I questioned/mentioned that above.
> Re why we don't do it at client side - @alex-zaitsev could you please comment on this?

We need it for HTTP queries in order to see the query time w/o network latency.

**Pull request closed without merging**